File: solutions/leetcode_198.c
int helper(int* nums, int numsSize, int16_t cache[101])
{
    if (numsSize <= 0) return 0;
    if (cache[numsSize] != -1) return cache[numsSize];

    cache[numsSize] = nums[numsSize - 1] + helper(nums, numsSize - 2, cache);
    int a = helper(nums, numsSize - 1, cache);
    if (a > cache[numsSize])
        cache[numsSize] = a;

    return cache[numsSize];
}

int rob(int* nums, int numsSize)
{
    int16_t cache[101];
    memset(cache, -1, 101 * sizeof(int16_t));
    return helper(nums, numsSize, cache);
}

File: solutions/leetcode_198.py
from functools import cache


class Solution:
    def rob(self, nums: list[int]) -> int:
        @cache
        def helper(start):
            if start >= len(nums):
                return 0
            return max(
                nums[start] + helper(start + 2),
                helper(start + 1)
            )
        return helper(0)


if __name__ == '__main__':
    given_nums = [2, 7, 9, 3, 1]
    expected = 12

    solution = Solution()
    result = solution.rob(given_nums)
    assert result == expected
